src/box/journal.c

```diff
@@ -94,36 +93,73 @@ journal_entry_fiber_wakeup_cb(struct journal_entry *entry)
 void
 journal_queue_wakeup(void)
 {
-	struct rlist *list = &journal_queue.waiters;
-	if (!rlist_empty(list) && !journal_queue_is_full())
-		fiber_wakeup(rlist_first_entry(list, struct fiber, state));
+	if (!stailq_empty(&journal_queue.requests) &&
+	    !journal_queue_is_full()) {
+		struct journal_entry *req =
+				stailq_first_entry(&journal_queue.requests,
+						   typeof(*req), fifo);
+		fiber_wakeup(req->fiber);
+	}
 }
 
-void
-journal_queue_wait(void)
+int
+journal_queue_wait(struct journal_entry *entry)
 {
-	if (!journal_queue_is_full() && !journal_queue_has_waiters())
-		return;
-	++journal_queue.waiter_count;
-	rlist_add_tail_entry(&journal_queue.waiters, fiber(), state);
-	/*
-	 * Is woken up when this position in the queue should go into the next
-	 * journal batch.
-	 */
+	if (!journal_queue_is_full() &&
+	    stailq_empty(&journal_queue.requests))
+		return 0;
+	int rc = -1;
+	struct journal_entry *prev_entry =
+			stailq_last_entry(&journal_queue.requests,
+					  typeof(*prev_entry), fifo);
+	stailq_add_tail_entry(&journal_queue.requests, entry, fifo);
+	assert(entry->fiber == NULL);
+	entry->fiber = fiber();
 	fiber_yield();
-	--journal_queue.waiter_count;
-	journal_queue_wakeup();
+	if (entry->is_complete) {
+		/* Already rolled back on cascade rollback. */
+		diag_set_journal_res(entry->res);
+	} else if (fiber_is_cancelled()) {
```

Comment:
Can't we just ignore cancellation while waiting for the journal queue? AFAICS we ignore it while waiting for WAL, i.e. this fix makes the behavior of `box.commit()` inconsistent:

https://github.com/tarantool/tarantool/blob/413e4e2bb98cace8a4ed10b9d41043ffff4bc0ce/src/box/txn.c#L1192-L1193

Reply:
Discussed f2f. Our current vision is to make fibers cancellable, we have a related task #9721. Also we should not continue operation and finish fiber when fiber is cancelled because our waiting primitives may enter busy loop.

Mentioned this place in #9721.